HOL_To_IMP_Minus/Compile_Nat/compile_nat.ML

```diff
@@ -176,23 +189,30 @@ fun IMP_Minus_of_tc_ast (f_ret_name, f_arg_names) lookup_fn register_sep e =
         in
           mk_Seq (es_IMP @ [\<^Const>\<open>tTAIL\<close>])
         end
-      | go bounds cont (Call (_, ((g_name, _), es))) =
+      | go bounds cont (Call (idx, ((g_name, _), es))) =
         let
           val g_compiled = lookup_fn g_name
+          val _ = List.length (#arg_regs g_compiled) = List.length es
+            orelse raise PARTIAL ("Call must occur fully applied", g_name)
           val g_ret = #ret_reg g_compiled
-          fun g_arg_reg i = nth (#arg_regs g_compiled) i |> HOLogic.mk_string
-          fun g_arg_cont i rhs = \<^Const>\<open>tAssign for \<open>g_arg_reg i\<close> rhs\<close>
-          val g_args_IMP =
-            map_index (fn (i, e) => go bounds (g_arg_cont i) e) es
+          fun g_local_arg_reg i = mk_local_reg ["Call", Int.toString idx, (* g_name, problem(?): can contain register_sep *) Int.toString i]
```

Comment:
Adding `g_name` to the argument name could increase readability of generated code, but I'm worried about name clashes